Mathlib/Geometry/RingedSpace/LocallyRingedSpace/ResidueField.lean

```diff
@@ -0,0 +1,122 @@
+/-
+Copyright (c) 2024 Christian Merten. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Christian Merten
+-/
+import Mathlib.Geometry.RingedSpace.LocallyRingedSpace
+
+/-!
+
+# Residue fields of points
+
+Any point `x` of a locally ringed space `X` comes with a natural residue field, namely the residue
+field of the stalk at `x`. Moreover, for every open subset of `X` containing `x`, we have a
+canonical evaluation map from `Γ(X, U)` to the residue field of `X` at `x`.
+
+## Main definitions
+
+The following are in the `AlgebraicGeometry.LocallyRingedSpace` namespace:
+
+- `residueField`: the residue field of the stalk at `x`.
+- `evaluation`: for open subsets `U` of `X` containing `x`, the evaluation map from sections over
+  `U` to the residue field at `x`.
+- `evaluationMap`: a morphism of locally ringed spaces induces a morphism, i.e. extension, of
+  residue fields.
+
+-/
+
+universe u
+
+open CategoryTheory TopologicalSpace Opposite
+
+noncomputable section
+
+namespace AlgebraicGeometry.LocallyRingedSpace
+
+variable (X : LocallyRingedSpace.{u}) {U : Opens X}
+
+/-- The residue field of `X` at a point `x` is the residue field of the stalk of `X`
+at `x`. -/
+def residueField (x : X) : CommRingCat :=
+  CommRingCat.of <| LocalRing.ResidueField (X.stalk x)
+
+instance (x : X) : Field (X.residueField x) :=
+  inferInstanceAs <| Field (LocalRing.ResidueField (X.stalk x))
+
+/--
+If `U` is an open of `X` containing `x`, we have a canonical ring map from the sections
+over `U` to the residue field of `x`.
+
+If we interpret sections over `U` as functions of `X` defined on `U`, then this ring map
+corresponds to evaluation at `x`.
+-/
+def evaluation (x : U) :
+    X.presheaf.obj (op U) ⟶ X.residueField x :=
+  X.presheaf.germ x ≫ LocalRing.residue _
+
+lemma evaluation_eq_zero_iff_mem_maximalIdeal (x : U) (f : X.presheaf.obj (op U)) :
+    X.evaluation x f = 0 ↔ (X.presheaf.germ x) f ∈ LocalRing.maximalIdeal (X.stalk x) :=
+  LocalRing.residue_eq_zero_iff _
+
+lemma evaluation_ne_zero_iff_isUnit {U : Opens X} (x : U) (f : X.presheaf.obj (op U)) :
+    X.evaluation x f ≠ 0 ↔ IsUnit ((X.presheaf.germ x) f) :=
+  LocalRing.residue_ne_zero_iff_isUnit _
+
+/-- The global evaluation map from `Γ(X, ⊤)` to the residue field at `x`. -/
+def Γevaluation (x : X) : LocallyRingedSpace.Γ.obj (op X) ⟶ X.residueField x :=
+  X.evaluation ⟨x, show x ∈ ⊤ from trivial⟩
+
+lemma Γevaluation_eq_zero_iff_mem_maximalIdeal (x : X) (f : LocallyRingedSpace.Γ.obj (op X)) :
+    X.Γevaluation x f = 0 ↔
+      (X.presheaf.germ ⟨x, show x ∈ ⊤ by trivial⟩) f ∈ LocalRing.maximalIdeal (X.stalk x) :=
+  LocalRing.residue_eq_zero_iff _
+
+lemma Γevaluation_ne_zero_iff_isUnit (x : U) (f : X.presheaf.obj (op U)) :
+    X.evaluation x f ≠ 0 ↔ IsUnit ((X.presheaf.germ x) f) :=
+  LocalRing.residue_ne_zero_iff_isUnit _
+
+@[simp]
+lemma evaluation_ne_zero_iff_mem_basicOpen (f : X.presheaf.obj (op U)) (x : U) :
```

Comment:
maybe `evaluation_eq_zero_iff_not_mem_basicOpen` as simp instead.

Reply:
Instead or additionally?

Reply:
I thought lint was complaining that the LHS is not simpNF as `ne_eq` rewrites it to `\not _ = _`

Reply:
No, it's only complaining for the `Γ` version, because it rewrites with `Γ_obj`.

Reply:
Ah okay. `evaluation_eq_zero_iff_not_mem_basicOpen` can still be added as simp lemma though. And then this lemma would be provable by simp.

Reply:
Should I still keep the `evaluation_ne_zero_iff_mem_basicOpen` version?

Reply:
I kind of believe that `evaluation_eq_zero_iff_not_mem_basicOpen` and `evaluation_ne_zero_iff_mem_basicOpen` (and their `Γ`-counterparts) are the only `evaluation_*_zero_iff_*` lemmas worth keeping here, since `mem_basicOpen` should currently be our simpNF of this concept.

Reply:
Not 100% sure though. Will need to use this API more to find out, but this is my gut feeling.

Reply:
I _thought_ I did use the `mem_maximalIdeal` versions somewhere in my application, but turns out I did not. So I am happy to remove them.

Reply:
Even if you did, I believe you should do that in two rewrites (or `A.trans B.symm`). It is more manageable for both mathlib and users if there is one canonical way to spell everything. This leads to n rewrite lemmas for n equivalent spellings instead of n^2, and the user will then always be encouraged to change their arguments into the canonical form, reliably knowing that this where all the API is.

Reply:
Understood.